core/ledger/kvledger/txmgmt/privacyenabledstate/common_storage_db.go

```diff
@@ -305,34 +318,33 @@ func (s *CommonStorageDB) HandleChaincodeDeploy(chaincodeDefinition *cceventmgmt
 		return nil
 	}
 
-	for directoryPath, indexFiles := range dbArtifacts {
-		// split the directory name
-		directoryPathArray := strings.Split(directoryPath, "/")
-		// process the indexes for the chain
+	for directoryPath, indexFiles := range dbIndexArtifacts {
 		indexFilesData := make(map[string][]byte)
 		for _, f := range indexFiles {
 			indexFilesData[f.FileHeader.Name] = f.FileContent
 		}
-		if directoryPathArray[3] == "indexes" {
+
+		if hasIndexForChaincode(directoryPath) {
 			err := indexCapable.ProcessIndexesForChaincodeDeploy(chaincodeDefinition.Name, indexFilesData)
 			if err != nil {
 				logger.Errorf("Error processing index for chaincode [%s]: %s", chaincodeDefinition.Name, err)
 			}
 			continue
 		}
-		// check for the indexes directory for the collection
-		if directoryPathArray[3] == "collections" && directoryPathArray[5] == "indexes" {
-			collectionName := directoryPathArray[4]
-			_, ok := collectionConfigMap[collectionName]
-			if !ok {
-				logger.Errorf("Error processing index for chaincode [%s]: cannot create an index for an undefined collection=[%s]", chaincodeDefinition.Name, collectionName)
-			} else {
-				err := indexCapable.ProcessIndexesForChaincodeDeploy(derivePvtDataNs(chaincodeDefinition.Name, collectionName),
-					indexFilesData)
-				if err != nil {
-					logger.Errorf("Error processing collection index for chaincode [%s]: %s", chaincodeDefinition.Name, err)
-				}
-			}
+
+		if !hasIndexForCollection(directoryPath) {
```

Comment:
You can have this function return collection name as well

Reply:
To maintain consistency between `hasIndexForChaincode` and `hasIndexForCollection`, I think it is okay to have a separate `getCollectionName` function. Though we might split the directory path thrice, I left it as-is given that deployment is an infrequent task.

Reply:
Replacing hard-coded numbers by `consts` is certainly good but beyond that, to be honest, I am not sure, if this PR improves much on the existing code.
A better way to make it more readable would be to contain the path parsing logic in function that simply returns a struct something like this...
`DBArtifactInfo{
    isIndexSpec bool,
    isCollectionLevel bool,
   collectionName string,
}` 
wdyt?

Reply:
The goal of this PR is just to replace the hardcoded number as it is a bad pattern as described in the PR description. This PR retains most of the existing logic as-is.

As I mentioned, I am not comfortable calling it `dbArtifacts`. Hence, I am not sure about using `isIndexSpec bool` either. We have only index artifacts for now.

I can do the following if there is an agreement.

(1) Rename the `dbArtifacts` to `dbIndexArtifacts`
(2) The following combines all three functions into a single one.
```
type struct indexInfo {
     hasIndexForChaincode bool;
     hasIndexForCollection bool;
     collectionName string;
}
func getIndexInfo(indexPath string) {
     indexInfo := &indexInfo{}
     dirsDepth := strings.Split(path, "/")
     if dirsDepth[chaincodeIndexDirDepth] == "indexes" {
           indexInfo.hasIndexForChaincode = true
           return indexInfo
     }
     if dirsDepth[collectionDirDepth] == "collections" &&
           dirsDepth[collectionIndexDirDepth] == "indexes" {
           indexInfo.hasIndexForCollection = true
           dirsDepth[collectionNameDepth]
     }
     return indexInfo     
}
```
The readability might be subjective. To me, having three functions called `hasIndexForChaincode()`, `hasIndexForCollection()`, and `getCollectionName()` was more readable than this. There is nothing wrong in having small functions. When we have a single function called `getIndexInfo()` combining all three, it has an additional info like if we have chaincode index, skip collection check. When we have three function, this context was only present in the `HandleChaincodeDeploy()`.

(3) The HandleChaincodeDeploy would look like (does not differ much from the current format)
```
indexInfo := getIndexInfo(directoryPath)
if indexInfo.hasIndexForChaincode {
     ....
     continue
}
if !indexInfo.hasIndexForCollection {
     ....
     continue
}
_, ok := collectionConfigMap[indexInfo.collectionName]
```

Reply:
> The goal of this PR is just to replace the hardcoded number as it is a bad pattern as described in the PR description.

If you would have stopped there, I had no problem. You missed my comment perhaps when I say "Replacing hard-coded numbers by consts is certainly good but beyond that.."

To understand a simple path structure, I have to move around three super tiny one-liner functions, not sure how it is better than what is already there. Also, parsing same string thrice, is more unnatural style to me. I surely, do not want to spend our time on these nit-picking like discussions. I think as a middle ground, you can just use const instead of hard coding.

Reply:
I followed **the most important concept first and the low-level details in the bottom**. Most of the time, we would not need to read low-level details. 

**Example-1:**
Assume that tomorrow, we introduce an index for `implicit_collection`. Though it can be still under the collection folder, I am taking this example to explain the reasoning. The developer needs to read only a few information and add a function called `hasIndexForImplicitCollection(directoryPath)` and even the `directoryPath` can be of very different format as compared to the existing path.

**Example-2:**
Suppose say some developer wants to read how index is deployed on CouchDB. Reading the `HandleChaincodeDeploy()` function alone is adequate and the developer can skip low-level details implemented in `hasIndexForChaincode()` `hasIndexForCollection()` `getCollectionName()`.

**Example-3:**
Assume that we change the `directoryPath` of collection. The developer has to simply go to `hasIndexForCollection()` and make the changes.

```
    if hasIndexForChaincode(directoryPath) {
    if !hasIndexForCollection(directoryPath) {
    collectionName := getCollectionName(directoryPath)
```
has better readability than the following IMO
```
    if directoryPathArray[chaincodeIndexDirDepth] == "indexes" {
    if directoryPathArray[collectionDirDepth] == "collections" && directoryPathArray[collectionIndexDirDepth] == "indexes" {
    collectionName := directoryPathArray[collectionNameDepth]
```
> Also, parsing same string thrice, is more unnatural style to me.

As it is an infrequent task, some amount of code duplication is fine.

Reply:
>  I surely, do not want to spend our time on these nit-picking like discussions. 

I agree and I am sorry for that. Given that I have strong opinion on **the most important concept first and the low-level details in the bottom**, I am trying to convince you.

Reply:
> I agree and I am sorry for that. Given that I have strong opinion on **the most important concept first and the low-level details in the bottom**, I am trying to convince you.
Anyone would hardly disagree to any philosophy of this sort in abstract. The difference of opinions arise from the situation you applying this to and at what granularity.
In this particular case, you have three functions (mostly one liners), they have to be invoked in exact same order (else they will panic). Moreover, this sequence of call is exactly at one place.

Reply:
In the initial PR, I did have a check on the len of the `dirDepth` to avoid this panic. Then, I thought that I would get a comment saying why are we changing the existing logic. 

I don't think we will come to an agreement on this. I am fine with closing this PR.

Reply:
Agreed on a middle-ground and updated the PR.